expand.go

```diff
@@ -0,0 +1,217 @@
+		if !f(buf[i]) {
+			return i
+		}
+	}
+	return -1
+}
+
+// Transform expands shell-like sequences like $foo and ${foo} using
+// the configured expand function.  The sequence '$$' is replaced with
+// a literal '$'.
+func (e *expandTransformer) Transform(dst, src []byte, atEOF bool) (int, int, error) {
+	var i int
+	var j int
+
+	for i < len(src) {
+
+		if j == len(dst) {
+			return j, i, transform.ErrShortDst
+		}
+
+		end := bytes.IndexByte(src[i:], '$')
+
+		if end == -1 {
+			// src does not contain '$', copy into dst
+			cnt := copy(dst[j:], src[i:])
+			i += cnt
+			j += cnt
+			continue
+		} else if end > 0 {
+			// copy chars preceding '$' from src to dst
+			cnt := copy(dst[j:], src[i:i+end])
+			i += cnt
+			j += cnt
+
+			if j == len(dst) {
+				return j, i, transform.ErrShortDst
+			}
+
+			// fallthrough
+		}
+
+		// src[i] now points to '$', j < len(dst)
+
+		// If we're at the end of src, but we found a starting
+		// token, return ErrShortSrc, unless we're also at EOF,
+		// in which case just copy it dst.
+		if i+1 == len(src) {
+			if atEOF {
+				dst[j] = src[i]
+				i++
+				j++
+				continue
+			}
+			return j, i, transform.ErrShortSrc
+		}
+
+		// At this point we know that src[i+1] is populated.
+
+		// If this token sequence represents the special '$$'
+		// sequence, emit a '$' into dst.
+		if src[i+1] == '$' {
+			dst[j] = src[i]
+			i += 2
+			j++
+			continue
+		}
+
+		var token []byte
+
+		// Start of bracketed token ${foo}
+		if src[i+1] == '{' {
+			end = bytes.IndexByte(src[i+2:], '}')
+			if end == -1 {
+				if atEOF {
+					// No closing bracket and we're at
+					// EOF, so it's not a valid bracket
+					// expression.
+					cnt := copy(dst[j:], src[i:])
+					if i+cnt == len(src) {
+						i += cnt
+						j += cnt
+						continue
+					}
+
+					// If dst didn't have enough space
+					// return ErrShortDst
+					return j, i, transform.ErrShortDst
+				}
+
+				// Otherwise, we need more bytes in src
+				return j, i, transform.ErrShortSrc
+			}
+
+			// Adjust end so it points to the byte
+			// immediately after the closing '}'
+			end += i + 3
+
+			token = src[i+2 : end-1]
+		} else { // Else start of non-bracketed token $foo
+			if !isShellNameFirstChar(src[i+1]) {
+				// If it doesn't conform to the naming
+				// rules for shell variables, do not
+				// try to expand, just copy to dst.
+				dst[j] = src[i]
+				i++
+				j++
+				continue
+			}
+
+			end = bytesIndexCFunc(src[i+2:], isShellNameChar)
+
+			if end == -1 {
+				// Reached the end of src without finding
+				// end of shell variable
+				if !atEOF {
+					// We need more bytes in src
+					return j, i, transform.ErrShortSrc
+				}
+				end = len(src)
+			} else {
+				// Adjust end so it points to the byte
+				// immediately after the token
+				end += i + 2
+			}
+
+			token = src[i+1 : end]
+		}
+
+		replacement, err := e.expand(string(token))
+		if err != nil {
+			return j, i, err
+		}
+
+		cnt := copy(dst[j:], replacement)
```

Comment:
Shouldn't you make sure there's enough room in `dst` before copying into it? If the replacement is too large, it'll be copied partially, then copied again when this is retried after `ErrShortDst` is returned.

Reply:
The copy itself was being used to check whether there was enough room.  I expect the normal case to be that there is enough room, but sure, we can the length and then avoid the unnecessary copy here and above.

Reply:
copy will copy at most `len(dst)` bytes, right? The issue I was pointing out was that in the case where `dst` is too short, copy will copy some number of bytes less than `len(src)`, we'll return `ErrDstShort`, in which case text/transform will retry after resizing dst, and we'll copy the part of `src` that was already copied again. (This all, assuming I understand the contract of text/transform correctly.)

Reply:
Yes that's correct.  I think I understood your original comment.  In my reply when I referred to "the unnecessary copy", I was talking about the case you described, the first copy that would happen when dst was too short, and which would then be performed again in full after dst was resized.